airflow/providers/apache/hive/hooks/hive.py

```diff
@@ -141,6 +141,7 @@ def _prepare_cli_cmd(self) -> list[Any]:
 
         if self.use_beeline:
             hive_bin = "beeline"
+            self._validate_beeline_parameters(conn)
```

Comment:
Would it make sense to push this validation to the constructor? The Connection object is retrieved there as well as determining the `use_beeline` attr. Could fail faster before even executing the task. WDYT?

Reply:
hmm, didn't notice maybe makes sense here since we have conn in `__init__`. 
Not related to changes in this PR but should we think of moving `self.get_connection(hive_cli_conn_id)` call outside `__init__` otherwise it will make a db call to get conn whenever dag parsing will happen?

Reply:
The db call would only happen if an operator using this hook constructed the hook in its `__init__`. There is _almost_ an instance of that in the [HiveOperator](https://github.com/apache/airflow/blob/28126c12fbdd2cac84e0fbcf2212154085aa5ed9/airflow/providers/apache/hive/operators/hive.py#L119), but it's to setup lazy loading the hook and make it an instance attr.

But yes, _ideally_ the constructors of hooks, operators, sensor, etc. don't make any external calls

Reply:
Additionally, If you initialize connection in __init__ you cannot make connection_id templated - even if you extend the operator.

While I used to think the same @josh-fell that early validation helps to surface such problems earlier, I think It has evolved and I am now of an opinion that constructors in our operators should **just** assign fields. Full stop.

And in the past (cannot find it easily) @uranusjr proposed that we get rid of explicit constructor and turn all our operators in 'dataclasses' https://peps.python.org/pep-0557/ - where you have only fields declared and all the __init__ boilerplate is generated.

And I tend to agree that would be a good idea.

Reply:
I agree that we need to avoid anything which might cause call to Airflow DB or any other resource. 
There is situation when hook initialised in operator not lazily and even expected as one of argument: 
[SSHHook](https://github.com/apache/airflow/blob/50b30e5b92808e91ad9b6b05189f560d58dd8152/airflow/providers/ssh/hooks/ssh.py#L143-L151) and [SSHOperator](https://github.com/apache/airflow/blob/50b30e5b92808e91ad9b6b05189f560d58dd8152/airflow/providers/ssh/operators/ssh.py#L81-L82)

Also I agree that operator should only allow fields declared however even if I'am a fan of dataclasses there is couple of issues exists:
1. [`__post_init__`](https://docs.python.org/3/library/dataclasses.html#post-init-processing), there will be temptation to use it.
2. A bit dumb inheritance, you should decorate child class with `@dataclass` otherwise it turned into the regular class with part of dataclass 🙄 
3. kwargs-only dataclasses introduced only in Python 3.10+

Reply:
>I am now of an opinion that constructors in our operators should just assign fields. Full stop.
Yes, same. I guess I figured since this was a hook, we're not moving the connection call out, and it's currently not going to get called in an operator, why not push up the validation? But I see it was misguided suggestion.

> ...get rid of explicit constructor and turn all our operators in 'dataclasses' 
I would be over the moon if this was the implementation or `attrs`; the latter has been life changing. It's a shame you can't build an operator now, with any real added value in simplication, with `attrs` because of the metaclass logic going on. Alas, I would love this.

Reply:
> Alas, I would love this.

I share the sentiment  :)